### ISSUE TITLE
investigate CDN performance

### ISSUE THREAD

**Issue by bridgetkromhout** [devopsdays/devopsdays-web#8190]
When I was in Cape Town last week, I noticed the devopsdays.org website was noticeably slower than one might expect if a CDN were helping. According to @adrianmoisey this was not an aberration and is in fact always that way. It's worth tracing through our CDN setup and investigating where the disconnect is.

**Comment by mattstratton** [comment]
Eh. Netlify is our CDN. Maybe they suck in Cape Town? There’s not really anything to configure for us unfortunately.

**Comment by adrianmoisey** [comment]
Looks to me like they're using AWS, and are sending us to Sao Paulo:
```
adrian@box:~ $ ping -c 5 devopsdays.org
PING devopsdays.org (52.67.148.165) 56(84) bytes of data.
64 bytes from ec2-52-67-148-165.sa-east-1.compute.amazonaws.com (52.67.148.165): icmp_seq=1 ttl=44 time=322 ms
64 bytes from ec2-52-67-148-165.sa-east-1.compute.amazonaws.com (52.67.148.165): icmp_seq=2 ttl=44 time=322 ms
64 bytes from ec2-52-67-148-165.sa-east-1.compute.amazonaws.com (52.67.148.165): icmp_seq=3 ttl=44 time=323 ms
64 bytes from ec2-52-67-148-165.sa-east-1.compute.amazonaws.com (52.67.148.165): icmp_seq=4 ttl=44 time=323 ms
64 bytes from ec2-52-67-148-165.sa-east-1.compute.amazonaws.com (52.67.148.165): icmp_seq=5 ttl=44 time=322 ms

--- devopsdays.org ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 5ms
rtt min/avg/max/mdev = 321.925/322.431/322.874/0.780 ms
```

I'm pretty sure it looked like I was hitting US East a few days ago.

We do have an AWS region being build in South Africa, it should go online in H1 of 2020. Hopefully Netlify can start using that.

I assume Devopdays won't change CDN?

Both Cloudflare and AWS have edges in South Africa (and seem to have global coverage)

For what it's worth, Netlify are using HTTP/2, which I think does help a little bit with how quickly the site loads.

**Comment by mattstratton** [comment]
As it happens, `legacy.devopsdays.org` and `assets.devopsdays.org` go through the CloudFlare CDN.

We don't use CloudFlare on `www` because of how Netlify does the build and caches things; if we used the CloudFlare CDN ostensibly we would be losing the cache invalidation that we get as part of having it on the same CDN network provided.

The thing when it comes to a lot of the stuff with how we build using Netlify is that we aren't building in our own infrastructure where we can write scripts to purge caches, etc. 

That being said, Netlify advertises themselves as being "world wide". Of course, you get "more global POP's" if you pay the $500+/month Enterprise plan.

I'm going to dig into this article and see if it can help us https://codewithhugo.com/enable-cdn-netlify/

**Comment by mattstratton** [comment]
I'm starting to believe we can likely push onto Cloudflare. I need to do a bit more thinking to see if anything breaks.

One thing that the linked blog article talks about is adding more caching to CSS/JS; we could do this, but the problem is they do change. The blog author handles this with fingerprinting those files; we used to do this in the build, but it results in every single html file having to be re-uploaded to Netlify on every build.

Currently since we aren't using the asset optimization feature of Netlify, it seems like it's not running those files through their main CDN (which is actually Cloudfront). I'm going to try turning the asset optimization back on and we can see if that helps. I'll ping @adrianmoisey 	once that is done to do another test.

**Comment by mattstratton** [comment]
Oof. Turning on the asset optimization (even without it actually processing the CSS/JS/image files) is making the build sloooooow. It could be because of the insecure links to the feed (which I have fixed in another PR), but I feel like we are gonna have to kill that feature again.

**Comment by mattstratton** [comment]
So the good news is that switching on the asset optimization makes the site performance amazing.

The bad news is, all builds now take 20+ minutes longer.

I'll have to do some more tweaking to see if I can fix that.

**Comment by adrianmoisey** [comment]
> So the good news is that switching on the asset optimization makes the site performance amazing.

From my tests:
```
adrianmoisey@Adrians-MacBookPro:~$ ping -c 5 d33wubrfki0l68.cloudfront.net
PING d33wubrfki0l68.cloudfront.net (143.204.65.103): 56 data bytes
64 bytes from 143.204.65.103: icmp_seq=0 ttl=246 time=3.588 ms
64 bytes from 143.204.65.103: icmp_seq=1 ttl=246 time=3.217 ms
64 bytes from 143.204.65.103: icmp_seq=2 ttl=246 time=4.026 ms
64 bytes from 143.204.65.103: icmp_seq=3 ttl=246 time=7.583 ms
64 bytes from 143.204.65.103: icmp_seq=4 ttl=246 time=3.086 ms

--- d33wubrfki0l68.cloudfront.net ping statistics ---
5 packets transmitted, 5 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 3.086/4.300/7.583/1.674 ms
```

Site also felt like it was running on my laptop, it was so fast ⚡️ 

> The bad news is, all builds now take 20+ minutes longer.

That's definitely not great, and I think build times is very important to keep low.

At least the problem has been identified! Good job!

**Comment by mattstratton** [comment]
One thing that could burn us if we decide to flip to using Cloudflare (which we could do) is that the way we have SSL set up, we might run into issues - the certificate is currently provisioned by Netlify, and I think we need to do some ninja work to switch to one from Cloudflare (not the end of the world if we do)

**Comment by mattstratton** [comment]
OK, SSL won't be an issue. There's more of an issue about how cache invalidation works, since Netlify is doing a bunch of magic for us. But TBH they weren't doing much magic with the images, etc, which is what most of the advantage of a CDN is. I'm tempted to try flipping to Cloudflare and see if that improves things.

**Comment by mattstratton** [comment]
Of course, the latest merge to master was really fast. I propose we watch this for a while and see how it goes.

![image](https://user-images.githubusercontent.com/1280666/64526378-beece880-d2c8-11e9-854a-ed1e652c1326.png)

**Comment by mattstratton** [comment]
If merges to master are better now (we'll see shortly), I put in #8227 to disable post-processing on deploy previews, which will help on the previews (we only need it on deploys from master anyway)

**Comment by mattstratton** [comment]
Yay! Merges to master are back to a good speed even when using the asset processing.

![image](https://user-images.githubusercontent.com/1280666/64527080-a382dd00-d2ca-11e9-9bc0-98b2471ccbc4.png)

**Comment by mattstratton** [comment]
@bridgetkromhout & @adrianmoisey 	if you feel like we have improved things markedly, we can probably close this issue now.

**Comment by bridgetkromhout** [comment]
I have seen some speedy builds finish. If @adrianmoisey can confirm that the site still seems okay for him, I'm fine with closing this.

**Comment by adrianmoisey** [comment]
💯 times better! It feels like a real website now!
From Cape Town, I'd consider the problem fixed.
Thanks all!
